xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -36,16 +39,39 @@
 public final class ServerWrapperForXds extends Server {
   private final Server delegate;
   private final XdsClientWrapperForServerSds xdsClientWrapperForServerSds;
+  @Nullable XdsServerBuilder.ErrorNotifier errorNotifier;
 
-  ServerWrapperForXds(Server delegate, XdsClientWrapperForServerSds xdsClientWrapperForServerSds) {
+  ServerWrapperForXds(
+      Server delegate,
+      XdsClientWrapperForServerSds xdsClientWrapperForServerSds,
+      @Nullable XdsServerBuilder.ErrorNotifier errorNotifier) {
     this.delegate = checkNotNull(delegate, "delegate");
     this.xdsClientWrapperForServerSds =
         checkNotNull(xdsClientWrapperForServerSds, "xdsClientWrapperForServerSds");
+    this.errorNotifier = errorNotifier;
   }
 
   @Override
-  public Server start() throws IOException {
-    delegate.start();
+  public Server start() {
+    xdsClientWrapperForServerSds.addServerWatcher(
+        new XdsClientWrapperForServerSds.ServerWatcher() {
+          @Override
+          public void onError(Status error) {
+            if (errorNotifier != null) {
+              errorNotifier.onError(error);
+            }
+          }
+
+          @Override
+          public void onSuccess(EnvoyServerProtoData.DownstreamTlsContext downstreamTlsContext) {
+            try {
+              delegate.start();
```

Comment:
If this throws an exception, it should cause start() itself to throw an exception. start() should block until delegate.start() completes.

Reply:
Oh okay. So to clarify:

- any exceptions from `delegate.start()` should just be thrown from this `start` without notifying via `errorNotifier`

- any error from the xdsClient will be notified via the `errorNotifier` but the `start` will remain blocked indefinitely (since there is no timeout) until a successful `onSuccess` is received from the xdsClient that will result in a `delegate.start()` call

Is my understanding correct?

Reply:
> any exceptions from delegate.start() should just be thrown from this start without notifying via errorNotifier

Yes. The notifier is for the xds watch which may take ~forever to complete.

> any error from the xdsClient will be notified via the errorNotifier but the start will remain blocked indefinitely (since there is no timeout) until a successful onSuccess is received from the xdsClient that will result in a delegate.start() call

Yes.

Reply:
Fixed.